zebra-state/src/in_memory/chains.rs

```diff
@@ -3,18 +3,97 @@ use crate::{Request, Response};
 use futures::prelude::*;
 use std::{
     pin::Pin,
+    sync::Arc,
     task::{Context, Poll},
 };
-use tower::Service;
+use tower::{Service, ServiceExt};
+use zebra_chain::{
+    block::{Block, BlockHeaderHash},
+    types::BlockHeight,
+};
+
+/// A representation of the chain state at a given height
+#[derive(Clone, Debug)]
+struct ChainState {
+    block: Arc<Block>,
+}
+
+/// A persistent data structure representing the end of a chain
+struct VolatileChain(im::OrdMap<BlockHeight, ChainState>);
+
+impl VolatileChain {
+    fn contains(&self, height: BlockHeight) -> bool {
+        todo!()
+    }
+}
 
 /// A service wrapper that tracks multiple chains, handles reorgs, and persists
 /// blocks to disk once they're past the reorg limit
 pub(crate) struct ChainsState<S> {
     /// The inner state service that only tracks a single chain
     inner: S,
+    /// The set of chains
+    //
+    // might need to use a map type and pop / reinsert with cummulative work as the index
```

Comment:
The consensus rule is: "the main tip has the highest cumulative work, breaking ties with the first block received by the node".

So I think zebra-consensus should manage the tip choice, but I'd like to get others' input.
```suggestion
    // might need to use a map type and pop / reinsert with cummulative work as the index
```

Reply:
Did you intend for the suggestion to make no changes?

Reply:
I fixed it. I think the alternatives are:
* (cumulative work: u64, timestamp: i64)
* is_main_tip: bool
* store the tips as a separate array, and the main tip as a reference, and make the cumulative work and timestamp (?) part of the chain state struct

It's impossible for the ancestor of a tip to have a higher cumulative work than the tip. So I don't think we need to track all the recent blocks as if they are tips.

Reply:
(I added some more design info.)

Reply:
@teor I don't know what you mean by "So I don't think we need to track all the recent blocks as if they are tips."

Reply:
I'm not sure if we're thinking about the same design.

How many items are in this set?
Just a few, or about 100?

Is it tracking just the tips, or all the recent blocks?

Reply:
its tracking all recent blocks, and it's approximately 100 but in practice probably larger because I plan on only removing blocks from the chains when they're persisted to disk, so all the blocks from chains that never get finalized will live on until the next restart.

Reply:
How is the inner state service used?
Which blocks does it have in it?
Why does there need to be a reference here?

Reply:
Oh, right, I'm really confused, because there is a ChainState and Chain**s**State. (Notice the extra "s".)

Let's rename them so it's clear what they do, and why they both exist.

Reply:
Oh, and there's a VolatileChain as well, which is "persistent".

Is there a comment somewhere that explains how these 3 data structures are used, and how they relate to each other?